architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+First, there are widely documented patterns and work-arounds for how to handle CRDs. The most common is the pattern of separating CRDs into a separate chart that then treats these CRDs as standard resources.
+
+Second, if that doesn't work for you, then there is a way to implement whatever system you want. Helm provides a plugin mechanism that can add new Helm commands.
+
+Third, Helm provides another mechanism that can send intermediate results to an external processor before issuing commands against the API server. This "post-render hook" system allows for custom processing in-flight.
+
+Finally, of course, Helm provides `helm template`, which allows users to render the chart into static YAML that they can then send to the cluster using their own tools.
+
+So there are _four separate avenues_ with which you can pursue implementing more advanced CRD handling. None of them require knowing Go or opening PRs or even getting feedback from the core maintainers. None of them require that you pass any of the quality gates enumerated above.
+
+You are free to handle CRDs however you want. To get you started, here are some useful links:
+
+- Read the [CRD Best Practices](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/) guide
+- Learn about [plugins](https://helm.sh/docs/topics/plugins/)
+- Use a [post-render](https://helm.sh/docs/topics/advanced/#post-rendering)
+- Export YAML with [helm template](https://helm.sh/docs/helm/helm_template/#helm-template)
+
+Helm maintainers are not trying to prevent usage of CRDs. As we've stated above, though, our quality bar is high because we have a lot of people depending on us. But we have provided ways for you to achieve your own goals without meeting our standards or objectives. And perhaps others would find use in the plugins or tools you provide.
+
+## Conclusion
+
+CRDs continue to present a huge challenge to Kubernetes users in general. Helm perhaps has it worse than other projects, as Helm is a generic solution to a generic problem (Kubernetes package management). Helm knows nothing of a cluster's stability, intent, architecture, or the social organization around it. Thus, to Helm, a development cluster with one user is no different than a thousand-node multi-tenant cluster with rigid RBACs.
+
+We earnestly want to find a solution to the CRD problem, but we are adamant about not sacrificing Helm's usability in the name of flexibility. If a simple CRD operation like an upgrade or a delete can destroy data across a cluster, or can silently break dozens of applications, we simply choose not to provide that operation via Helm.
+
+If we can solve the problems presented above, we can support CRDs robustly. That is our desired goal. But in our estimation, Kubernetes is not yet mature enough for us to be able to do this. When we begin work on Helm 4, we will of course revisit the subject, but the guidelines in this document will still serve as guidance on Helm 4.
```

Comment:
A CRD, being a global thing, makes a helm chart essentially have an external dependency on something. Helm was only really built with the intention that all dependencies are internal to a chart. And I think that's the real rub. There are other external dependency issues that helm has as well. CRD's are just the most commonly asked for thing. I think its reasonable to draw the line at internal only as there is a huge amount of complexity added when you support external dependencies as well, as we keep running into.

What may be in scope I think may be for helm to assist some other hypothetical CRD management tool, or sysadmin by giving it information it can use to help manage CRDs? So, kind of inline with something hinted above in the document about having the chart be able to declare what CRDs it depend on.

Reply:
> What may be in scope I think may be for helm to assist some other hypothetical CRD management tool, or sysadmin by giving it information it can use to help manage CRDs?

https://github.com/helm/helm/issues/8607 is an example of this direction.

Reply:
Ah, interesting. yeah. Perhaps something like that. But maybe one step further. Have the option to register the CRDS expected as part of the release but without loading them instead of just having them in the chart. Then the cluster could be introspected for releases and what CRDs they expect. For some charts, the user without CRD permissions could then install a release, and then ask the admin to go review the CRDS needed that are already waiting in the release but not installed yet. Then the hypothetical tool could load them from the release info into the cluster if it was "ok".